### PR TITLE
If the dataset is not exactly divisible by world_size, the validation accuracy is incorrect. We solve this problem with an auxiliary validation set.

### LINKED ISSUE #769: Add DistributedSampler for validation data loader in Imagenet example
In examples/imagenet/main.py, we only set a distributed sampler (i.e., `train_sampler = torch.utils.data.distributed.DistributedSampler(train_dataset)`) for `train_loader`, while neglecting setting the distributed sampler for `val_loader`. In this case, the loss and accuracy metrics of test logs are exactly the same among different GPUs as follows, leading to unnecessary computation.
How about adding DistributedSampler for both `train_sampler` and `val_loader`?

```
Test: [ 0/62]	Time 12.870 (12.870)	Loss 2.1374e+00 (2.1374e+00)	Acc@1   0.00 (  0.00)	Acc@5 100.00 (100.00)
Test: [10/62]	Time  0.461 ( 1.547)	Loss 2.1963e+00 (2.1630e+00)	Acc@1   0.00 (  0.00)	Acc@5  98.44 ( 99.72)
Test: [20/62]	Time  0.352 ( 0.969)	Loss 2.1709e+00 (2.1958e+00)	Acc@1   0.00 (  0.00)	Acc@5  98.44 ( 73.07)
Test: [30/62]	Time  0.487 ( 0.810)	Loss 2.3123e+00 (2.2959e+00)	Acc@1   0.00 (  0.00)	Acc@5  79.69 ( 62.75)
Test: [40/62]	Time  0.345 ( 0.714)	Loss 2.8967e+00 (2.3489e+00)	Acc@1   0.00 (  0.11)	Acc@5   1.56 ( 60.90)
Test: [50/62]	Time  0.485 ( 0.666)	Loss 2.4475e+00 (2.8809e+00)	Acc@1   0.00 ( 12.56)	Acc@5   0.00 ( 61.61)
Test: [60/62]	Time  0.401 ( 0.620)	Loss 2.2804e+00 (2.9429e+00)	Acc@1  18.75 ( 11.96)	Acc@5  21.88 ( 53.18)
 * Acc@1 11.924 Acc@5 52.917
Test: [ 0/62]	Time 12.978 (12.978)	Loss 2.1374e+00 (2.1374e+00)	Acc@1   0.00 (  0.00)	Acc@5 100.00 (100.00)
Test: [10/62]	Time  0.297 ( 1.554)	Loss 2.1963e+00 (2.1630e+00)	Acc@1   0.00 (  0.00)	Acc@5  98.44 ( 99.72)
Test: [20/62]	Time  0.150 ( 0.970)	Loss 2.1709e+00 (2.1958e+00)	Acc@1   0.00 (  0.00)	Acc@5  98.44 ( 73.07)
Test: [30/62]	Time  0.267 ( 0.810)	Loss 2.3123e+00 (2.2959e+00)	Acc@1   0.00 (  0.00)	Acc@5  79.69 ( 62.75)
Test: [40/62]	Time  0.145 ( 0.715)	Loss 2.8967e+00 (2.3489e+00)	Acc@1   0.00 (  0.11)	Acc@5   1.56 ( 60.90)
Test: [50/62]	Time  0.247 ( 0.666)	Loss 2.4475e+00 (2.8809e+00)	Acc@1   0.00 ( 12.56)	Acc@5   0.00 ( 61.61)
Test: [60/62]	Time  0.164 ( 0.620)	Loss 2.2804e+00 (2.9429e+00)	Acc@1  18.75 ( 11.96)	Acc@5  21.88 ( 53.18)
 * Acc@1 11.924 Acc@5 52.917
```

### PR DESCRIPTION
**Pull request by siaimes**
If the dataset is not exactly divisible by world_size, the validation accuracy is incorrect. We solve this problem with an auxiliary validation set.

 fix #769 close #907 #890

Signed-off-by: siaimes <34199488+siaimes@users.noreply.github.com>

Example results:

image
image
image
image
image

Note: ceil((50000 // 3) / int(256/3)) + 1 = 198 for 3 GPUs.

### CODE REVIEW

**Comment by msaroufim** [comment]
3 fixes in one, wow! Will need some test to test but tagging @mrshenli in case he can code review at a glance

**Review by cbalioglu — APPROVED** [review]
Thanks for the fix! Looks good to me!
